### ISSUE TITLE
Can we improve on the hardcoded `arg` name in this error?

### ISSUE THREAD

**Issue by DavisVaughan** [r-lib/vctrs#1581]
```r
vctrs::vec_cast(1, "x", x_arg = 1)
#> Error in `vctrs::vec_cast()`:
#> ! `arg` must be a string.
```

Ideally it would reference `x_arg` here.

This happens because `arg` is hardcoded here:
https://github.com/r-lib/vctrs/blob/79f5318b886a8316b594477a84d1dd8ab3373ece/src/arg.c#L136

But a lot of times `(*data).x` is a symbol with the right arg name, so maybe we can check to see if that is a symbol, and if so use `r_sym_c_string()` to turn it into a name we can use in the error?

**Comment by lionel-** [comment]
We should use `x_arg = caller_arg(x)` and thread that down on the C side using a lazy vctrs-arg.

**Comment by lionel-** [comment]
oh the problem was that `caller_arg()` may create non-string args. We should convert these to string with `as_label()`.

**Comment by DavisVaughan** [comment]
@lionel- I'm not sure that #1711 actually closes this issue.

This is purely about the fact that we should reference the correct argument name, `x_arg`, in the error message when `x_arg` itself is validated.

**Comment by DavisVaughan** [comment]
Oh wait, you think that `1` should be an allowed value for `x_arg`?? That seems so strange to me 😆 

```r
> vctrs::vec_cast(1, "x", x_arg = 1)
Error:
! Can't convert `1` <double> to <character>.
```

**Comment by lionel-** [comment]
yup we allow literals and any kind of R objects in error args. This way, checking functions do not fail when used interactively. Also it's useful to allow calls for some use cases, such as tibble errors in subset operators. Since operators like `[[` don't have argument names, we show the user input instead.
